### ISSUE TITLE
Application Gateway role assignment propagation time causes unauthorized Key Vault access

### ISSUE THREAD

**Issue by Xitric** [mspnp/aks-baseline#242]
#### Description
We have utilized the baseline for deploying an AKS cluster to our Azure subscription. Beyond migrating the templates to Bicep, we have made minor changes to the structure of the templates. However, we still believe that this issue is relevant for the baseline.

A recent PR (#241) has changed Key Vault access from the Application Gateway from using Access Policies to Azure RBAC, which we have also adopted in our version. However, this now causes the Application Gateway to intermittently fail a clean deployment, with a subsequent redeployment being successful.

#### Steps to reproduce
We have noticed that PR #241 adds a dependency from the application gateway resource to the private endpoint resource between the nodepool subnet and Azure key vault. This dependency seems completely arbitrary, since the application gateway runs in its own subnet and is not supposed to be dependent on nodepool access to key vault. Thus, we have removed that dependency.

1. Remove dependency between application gateway and the private endpoint `nodepools-to-akv`
2. Deploy the hub and spoke resources, as usual
3. Deploy the cluster stamp template to an empty resource group

The failed deployment is intermittent, and only happens in rare occasions (around 25% in my experience).

#### Expected result
The deployment completes successfully.

#### Actual result
On the first deployment to a clean resource group, we sometimes get the following conflict (some information is redacted to avoid accidentally exposing information):

```json
{
    "status": "Failed",
    "error": {
        "code": "ApplicationGatewayKeyVaultSecretException",
        "message": "Problem occured while accessing and validating KeyVault Secrets associated with Application Gateway '/subscriptions/<subid>/resourceGroups/rg-redacted/providers/Microsoft.Network/applicationGateways/agw-dev-westeurope'. See details below:",
        "details": [
            {
                "code": "0",
                "message": "Caller is not authorized to perform action on resource.\r\nIf role assignments, deny assignments or role definitions were changed recently, please observe propagation time.\r\nCaller: name=Microsoft.Network/applicationGateways;appid=<appid>;oid=<oid>;iss=https://sts.windows.net/<id>/\r\nAction: 'Microsoft.KeyVault/vaults/secrets/getSecret/action'\r\nResource: '/subscriptions/<subid>/resourcegroups/rg-redacted/providers/microsoft.keyvault/vaults/kv-aks-ftdypjm3o6ewm/secrets/ingress-ssl-certificate'\r\nAssignment: (not found)\r\nVault: kv-aks-ftdypjm3o6ewm;location=westeurope\r\n"
            },
            {
                "code": "0",
                "message": "Caller is not authorized to perform action on resource.\r\nIf role assignments, deny assignments or role definitions were changed recently, please observe propagation time.\r\nCaller: name=Microsoft.Network/applicationGateways;appid=<appid>;oid=<oid>;iss=https://sts.windows.net/<id>/\r\nAction: 'Microsoft.KeyVault/vaults/secrets/getSecret/action'\r\nResource: '/subscriptions/<subid>/resourcegroups/rg-redacted/providers/microsoft.keyvault/vaults/kv-aks-ftdypjm3o6ewm/secrets/gateway-ssl-certificate'\r\nAssignment: (not found)\r\nVault: kv-aks-ftdypjm3o6ewm;location=westeurope\r\n"
            }
        ]
    }
}
```

From my understanding, the Application Gateway creation was attempted so soon, that the role assignments had not been fully propagated, causing the unauthorized access attempt to key vault. This is despite our application gateway having explicit dependencies on both role assignment resources:

```json
"dependsOn": [
    "[resourceId('Microsoft.KeyVault/vaults/providers/roleAssignments', variables('keyVaultName'), 'Microsoft.Authorization', guid(resourceGroup().id, 'mi-appgateway-frontend', variables('keyVaultReader')))]",
    "[resourceId('Microsoft.KeyVault/vaults/providers/roleAssignments', variables('keyVaultName'), 'Microsoft.Authorization', guid(resourceGroup().id, 'mi-appgateway-frontend', variables('keyVaultSecretsUserRole')))]"
]
```

Our declarations in Bicep:

```bicep
dependsOn: [
  gatewayKeyVaultReaderRoleAssignment
  gatewayKeyVaultSecretUserRoleAssignment
]
```

#### Additional information
We wonder if the dependency on the private endpoint acts as an arbitrary "timer" delaying deployments of the application gateway long enough for the role assignments to be propagated. If this is the case, that implementation seems rather flaky, and a better solution should be found.

If there is another good reason why the dependency exists on the private endpoint `nodepools-to-akv` we would like to know, and perhaps a comment should be added to the code to make this dependency clearer. It is not clear to us why that dependency was necessary to add as part of #241.

**Comment by Xitric** [comment]
I see, so the fact that it is being associated with the nodepool subnet is just an arbitrary decision, and merely serves to give it an address space from which to pick an IP address. So that explains the need for the dependency, but can it also explain why we are getting intermittent unauthorized errors? In case that the private endpoint is not available yet, I assume that the workload then attempts to contact the key vault via other means (a public endpoint perhaps), which is not allowed in this baseline. Unfortunately, I cannot test this until the morning.

In any case, I think the improvements you propose are definitely welcome, and it is something that we will be bringing into our Bicep port as well.

**Comment by ckittel** [comment]
We have not yet picked up the "dedicated subnet" work for the two private endpoints (ACR & AKV). Hoping to get to it in December.

**Comment by ckittel** [comment]
@teilmeier, just FYI on your dedicated subnet test that this issue spun into that as well.  No change in direction, just letting you know that when you open a PR for that, you can link against this issue.

**Comment by ckittel** [comment]
Being addressed by @teilmeier in #291

**Comment by ckittel** [comment]
Migration to subnet dedicated to the private endpoints is complete.  Closing this issue as it was converted to track that.  Thanks all!
